r/tools/winlibs.R

```diff
@@ -0,0 +1,9 @@
+# Download static arrow from rwinlib
```

Comment:
This seems weird to me as a non-R users. Can you explain to me what this is about and whether this is going to be permanent?

Reply:
I also find it weird to not have the ability to build the static Arrow library using endogenous (aka inside apache/arrow) scripts

Reply:
Yes for sure. This is to help us get the package in CRAN.

While CRAN supports compiling packages with native code, it has several restrictions. Therefore, it is a common practice to precompile binaries for CRAN in Window that are downloaded when the package is build and tested in the CRAN build machines.

Short term, I think we can start by compiling ourselves the `arrow` binaries, but long term, there is nothing preventing us from automating this process in the `arrow` codebase.

Out of curiosity, how does this work with `pip install` packages? Does https://pypi.org provide sufficient infrastructure to compile all `arrow` binaries from source? Or do `arrow` maintainers upload precompiled binaries? How are those binaries built? Are the `pip install` binaries built automatically from in the arrow project?

Reply:
On https://pypi.org we provide pre-compiled binaries that are built and uploaded on each Arrow release. We have build infrastructure in https://github.com/apache/arrow/tree/master/dev/tasks to build all the binary releases of Arrow directly using CI services (for Python and CentOS/Debian packages). `pip install` from the source package on PyPI is only working when you have the C++ binaries installed, we hope to provide at one time a source Python package that is self-contained but as >95% of all users use the binary packages, this is not a concern at the moment.

> Short term, I think we can start by compiling ourselves the arrow binaries, but long term, there is nothing preventing us from automating this process in the arrow codebase.

Yes, we should automate this for R/CRAN as we do it for Python. As far as I understand it currently, you would need a static `libarrow.a`, `libparquet.a` and possibly a `libarrow_bundled_thirdparty_dependencies.a` built with MinGW available that is then pulled by CRAN to build the final package? It would then be sufficient to update the build of these static libs on each release?

Reply:
Yes, the following libraries would be needed: 
- parquet 
- arrow
- thrift
- boost_regex-mt-s 
- double-conversion

To automate building the R binaries in the `arrow` repo we would have to mimic the following repos (for the packages we need, not all):
- https://github.com/r-windows/rtools-packages
- https://github.com/r-windows/rtools-backports

It's probably not a ton of work since they already contain AppVeyor scripts; however, we might hit additional issues we are currently not aware of while publishing to CRAN. So I would suggest we focus on publishing in CRAN and then start automating additional build components as needed.

Reply:
BTW I'm a bit lost as to how getting on CRAN is going to be possible without Linux packages in the upstream package managers. This hasn't been explained anywhere

Reply:
Yes for sure, CRAN requires two mayor platforms, which we now support: OS X and Windows. Here is the reference:

> Package authors should make all reasonable efforts to provide cross-platform portable code. Packages will not normally be accepted that do not run on at least two of the major R platforms. Cases for Windows-only packages will be considered, but CRAN may not be the most appropriate place to host them.

https://cran.r-project.org/web/packages/policies.html